src/LuaTransformer.ts

```diff
@@ -3043,44 +3065,283 @@ export class LuaTransformer {
-                    const visitResults = expression.right.elements.map(e => this.transformExpression(e));
-                    right = this.filterUndefined(visitResults);
-                } else {
-                    right = [tstl.createNilLiteral()];
+            if (flattenable) {
+                const expressionType = this.checker.getTypeAtLocation(expression.right);
+                let right = this.transformExpression(expression.right);
+
+                if (
+                    !tsHelper.isTupleReturnCall(expression.right, this.checker) &&
+                    tsHelper.isArrayType(expressionType, this.checker, this.program)
+                ) {
+                    right = this.createUnpackCall(right, expression.right);
                 }
-            } else if (tsHelper.isTupleReturnCall(expression.right, this.checker)) {
-                right = [this.transformExpression(expression.right)];
-            } else {
-                right = [this.createUnpackCall(this.transformExpression(expression.right), expression.right)];
+
+                return this.transformFlattenableDestructuringAssignment(
+                    expression as ts.DestructuringAssignment,
+                    right
+                );
             }
-            return tstl.createAssignmentStatement(left as tstl.AssignmentLeftHandSideExpression[], right, expression);
+
+            let right = this.transformExpression(expression.right);
+            const rootIdentifier = tstl.createAnonymousIdentifier(expression.left);
+
+            if (tsHelper.isTupleReturnCall(expression.right, this.checker)) {
+                right = this.wrapInTable(right);
+            }
+
+            const rootDeclaration = tstl.createVariableDeclarationStatement(rootIdentifier, right);
+
+            const statements = this.transformDestructuringAssignment(
+                expression as ts.DestructuringAssignment,
+                rootIdentifier
+            );
+            statements.unshift(rootDeclaration);
+
+            return statements;
         } else {
             // Simple assignment
             return this.transformAssignment(expression.left, this.transformExpression(expression.right));
         }
     }
 
+    protected transformFlattenableDestructuringAssignment(
+        node: ts.DestructuringAssignment,
+        right: tstl.Expression | tstl.Expression[]
+    ): tstl.Statement {
+        if (ts.isArrayLiteralExpression(node.left)) {
+            const left: tstl.AssignmentLeftHandSideExpression[] = node.left.elements.map(
+                element => this.transformExpression(element) as tstl.AssignmentLeftHandSideExpression
+            );
+            return tstl.createAssignmentStatement(left, right, node);
+        }
+
+        throw TSTLErrors.NonFlattenableDestructure(node);
+    }
+
+    protected transformDestructuringAssignment(
+        node: ts.DestructuringAssignment,
+        root: tstl.Expression
+    ): tstl.Statement[] {
```

Comment:
Will need to discuss with @ark120202 and @tomblind if they are okay with this since this seems a major workaround. `ts.DestructuringAssignment` is a type of expression, so ideally this would be a public transform visitor returning a ExpressionVisitResult. I do however understand that without the preceding statements rewrite this might be pretty much impossible.